File: ios/Application/OpenArchiveUseCase.swift
import Foundation

final class OpenArchiveUseCase {
  private let engine: ArchiveEngine

  init(engine: ArchiveEngine) {
    self.engine = engine
  }

  func execute(input: ArchiveInput) async throws -> (inspection: ArchiveInspection, session: ArchiveEngineSession) {
    let archiveData = try await input.openForRandomAccess()
    let inspection = try engine.inspect(data: archiveData)
    let session = try ArchiveInMemorySession(entries: inspection.entries, archiveData: archiveData)
    return (inspection, session)
  }
}

final class ArchiveInMemorySession: ArchiveEngineSession {
  private let entriesByPath: [String: EntryDescriptor]
  let entries: [EntryDescriptor]
  let archiveData: Data?
  let archiveURL: URL?
  private let tempDir: URL?

  init(entries: [EntryDescriptor], archiveData: Data?) throws {
    self.entries = entries
    self.archiveData = archiveData
    if let archiveData {
      let tempDir = FileManager.default.temporaryDirectory
        .appendingPathComponent(UUID().uuidString)
        .appendingPathComponent("session")
      try FileManager.default.createDirectory(at: tempDir, withIntermediateDirectories: true)
      let archiveURL = tempDir.appendingPathComponent("archive.zip")
      try archiveData.write(to: archiveURL)
      self.tempDir = tempDir
      self.archiveURL = archiveURL
    } else {
      self.tempDir = nil
      self.archiveURL = nil
    }
    var map: [String: EntryDescriptor] = [:]
    for entry in entries {
      map[entry.path] = entry
    }
    self.entriesByPath = map
  }

  func entry(at path: String) -> EntryDescriptor? {
    entriesByPath[path]
  }

  deinit {
    if let tempDir {
      try? FileManager.default.removeItem(at: tempDir)
    }
  }
}

File: android/src/main/java/com/mcodex/nitroarchive/application/OpenArchiveUseCase.kt
package com.mcodex.nitroarchive.application

import com.mcodex.nitroarchive.application.ports.*

data class OpenArchiveResult(
    val inspection: ArchiveInspection,
    val session: ArchiveEngineSession
)

class OpenArchiveUseCase(private val engine: ArchiveEngine) {

    suspend fun execute(input: ArchiveInput): OpenArchiveResult {
        val inspection = engine.inspect(input)
        val data = input.openForRandomAccess()
        val session = object : ArchiveEngineSession {
            override val entries: List<EntryDescriptor> = inspection.entries
            override val archiveData: ByteArray? = data

            override fun entry(at: String): EntryDescriptor? {
                return inspection.entries.find { it.path == at }
            }
        }
        return OpenArchiveResult(inspection = inspection, session = session)
    }
}
